Cached `libclang-bindings` may not match current configuration

I did not upgrade my development system for quite some time, just to make sure that I did not break anything before ICFP and Poland.  (I miss NixOS generations!)  I finally upgraded this weekend, and the upgrade changed my default LLVM/Clang from version 20 to 21.  (Clang/LLVM 20 is no longer available, until I install the packages for it.)

Attempts to build `hs-bindgen` failed with the following error:

```
libclang.so.20.1: cannot open shared object file: No such file or directory
```

Since our `libclang-bindings` package is now in a separate project, build artifacts are stored in the Cabal store (in `~/.cabal/store`), *not* locally (in `dist-newstyle`).  The version of `libclang.so` that is linked when we build our `libclang-bindings` package is not reflected in the package hash.  The version of `libclang.so` that is linked when first building for a given GHC version (and revision of the code) is therefore always used, regardless of any environmental changes that configure a different LLVM/Clang version.

In this case, I upgraded from LLVM/Clang 20 to 21.  My environment was configured to use LLVM/Clang 21, but old builds of our `libclang-bindings` package still linked to `libclang.so.20.1`, which no longer exists on my system.

I have previously tested against the many version of LLVM/Clang available on my system by simply adjusting environment variables.  Now that our `libclang-bindings` package is in a separate project, this no longer works correctly.  The configuration only affects the current project: when working on `hs-bindgen`, the packages in `hs-bindgen` are built with the current configuration, but the cached (stored) `libclang-bindings` dependency is used regardless of any configuration changes.

`pkg-config` dependency configuration is reflected in a package hash, but we do not use `pkg-config` to configure LLVM/Clang.  Might there be another way to make the package hash reflect the LLVM/Clang version?

I think that this affects anybody who uses Cabal normally.  (Nix builds that do not use a persistent Cabal store should not be affected, as dependency configuration is a proper input and not just global state.)

If we are unable to make the `libclang-bindings` package hash reflect the FFI dependencies, perhaps we should document this, telling users how to resolve issues.  My understanding, however, is that there is no tooling available for removing specific build artifacts from the Cabal store.  One can manually remove stuff, but this makes the `package.cache` invalid and could cause problems.  Is removing the whole store (for each GHC version that contains a problematic artifact) the only "safe" option?

As for testing with multiple versions of LLVM/Clang during development, perhaps all such tests need to be done in environments that do not cache previous builds of `libclang-bindings` (that use a different version).

The reason why this has come up again is that users are going to start running into this soon now that LLVM/Clang 22 is released.

A change of LLVM/Clang installed on the system can result in a number of symptoms:

* A minor upgrade of LLVM/Clang is generally not fatal.  Users should see a version mismatch warning like the following:

    ```
    [Warning] [HsBindgen] [builtin-include-dir] clang version mismatch:
      libclang version: clang version 22.1.1
      clang version:    clang version 22.1.2
    ```

* A major upgrade of LLVM/Clang where the old version is removed from the system is fatal.  Users cannot run `hs-bindgen` at all:

    ```
    libclang.so.21.1: cannot open shared object file: No such file or directory
    ```

* A major upgrade of LLVM/Clang where the old version is not removed from the system causes issues in cases where a header from the compiler builtin include directory is used.

    ```
    [6 of 8] Compiling Test.TH.Test01   ( test/th/Test/TH/Test01.hs, .../Test01.o )
    Template Haskell error: [Error  ] [Libclang ] [clang] /usr/include/stdlib.h:32:10: fatal error: 'stddef.h' file not found

    Call to 'libclang' returned an error
    test/th/Test/TH/Test01.hs:1:1: error: [GHC-87897]
        Exception when trying to run compile-time code:
          ExitFailure 2
        Code: let
                cfg :: Config
                cfg = ...
                ....
              in withHsBindgen cfg cfgTh $ hashInclude "test_01.h"
      |
    1 | {-# LANGUAGE CApiFFI #-}
      | ^
    ```

  Note that this example is of a failure when building our tests when running `cabal build all`.

I will add documentation to the manual, but we are also trying to determine a better way to deal with the situation.

--- a/cabal.project.base
+++ b/cabal.project.base
@@ -5,7 +5,7 @@ index-state: 2026-03-12T10:22:50Z
 source-repository-package
     type: git
     location: https://github.com/well-typed/libclang
-    tag: 1054474fae403bfb52c7919680cac03d3d3d4237
+    tag: cdbf817187a261ebf8c31b32c85c5693eedf298d
 
 --
 -- Temporary package bounds overrides for ghc 9.14

--- a/flake.nix
+++ b/flake.nix
@@ -5,7 +5,7 @@
     flake-parts.url = "github:hercules-ci/flake-parts";
     nixpkgs.url = "github:NixOS/nixpkgs/nixpkgs-unstable";
     libclang-bindings-src = {
-      url = "github:well-typed/libclang?rev=1054474fae403bfb52c7919680cac03d3d3d4237";
+      url = "github:well-typed/libclang?rev=cdbf817187a261ebf8c31b32c85c5693eedf298d";
       flake = false;
     };
   };

--- a/manual/Installation.md
+++ b/manual/Installation.md
@@ -111,6 +111,84 @@ should be set in the `.cabal` file.
   This is only needed when you want to use a version of LLVM/Clang that is
   not in your current `PATH`.
 
+### Cabal store maintenance
+
+The `libclang` bindings used by `hs-bindgen` are implemented in package
+`libclang-bindings`.  When built, it is linked to the `libclang` and `libLLVM`
+shared objects for the version of LLVM/Clang that you are using, and the built
+`libclang-bindings` library is cached in the Cabal store.  Cabal does not
+consider such system dependencies to determine when a package needs to be
+rebuilt, however, so you might continue to use a library built with an older
+version of LLVM/Clang even after a system upgrade installs a newer version.
+
+Links to LLVM/Clang shared objects generally specify the minor version.  If an
+upgrade replaces an older version of LLVM/Clang with a newer version that has
+the same minor version (and a newer patch version), then the link in the
+cached `libclang-bindings` build is not broken.  The newer version *should*
+still work, though we recommend rebuilding `hs-bindgen`/`libclang-bindings`
+using the newer version.  In this case, `hs-bindgen` displays a warning like
+the following, where the "compile time version" is the version of LLVM/Clang
+used to to build the cached `libclang-bindings` library, and the "runtime
+version" is the version for the shared object that is dynamically linked at
+runtime.
+
+```
+[Warning] [HsBindgen] [boot] clang version mismatch:
+  clang compile time version: 22.1.2
+  clang runtime version:      22.1.3
+```
+
+If an upgrade installs a new major version and removes the older version, then
+the link in the cached `libclang-bindings` build is broken.  In this case, the
+dynamic library loader fails with an error like the following, preventing
+`hs-bindgen` from running at all.
+
+```
+libclang.so.20.1: cannot open shared object file: No such file or directory
+```
+
+Cabal does not provide an easy way to resolve this issue.  Note that this is an
+issue for any Haskell package that links to system libraries that do not have
+`pkg-config` support (in which case Cabal can track the dependency versions).
+It is particularly problematic with LLVM/Clang, however, because LLVM/Clang has
+frequent releases.  Currently, we suggest two ways to work around the issue.
+
+The easiest way to force Cabal to rebuild out-of-date/broken packages is to
+clear the cache.  You can determine your Cabal store directory by running
+`cabal path --store-dir`.  The Cabal store has separate caches for each version
+of GHC.  Recursively remove a GHC directory to clear the cache for that version.
+Note that deleting just a package directory is not advised because it can break
+the package database.  The downside to clearing your cache is that doing so may
+result in time-consuming recompilation of many other packages.
+
+Since we understand that clearing the cache might not be desirable, `hs-bindgen`
+offers a bespoke workaround, in the form of a compile-time setting specifying
+the LLVM/Clang version.  This forces Cabal to distinguish `libclang-bindings`
+builds that link to different versions, forcing a new build if one for the
+specified version is not already in the Cabal store.  Do this by adding the
+following to a `cabal.project.local` file for your project, with the desired
+version.
+
+```
+package libclang-bindings
+  ghc-options: -optc=-DCLANG_VERSION=22.1.3
+```
+
+When configured like this, `libclang-bindings` confirms that the version matches
+the version of LLVM/Clang used at compile time.  The following checks are
+supported:
+
+* `MAJOR` to just check the major version (example: `22`)
+* `MAJOR.MINOR` to check the major and minor versions (example: `22.1`)
+* `MAJOR.MINOR.PATCH` to check the major, minor, and patch versions
+  (example: `22.1.3`)
+
+Note that `ghc-options` specifying an `-optc` option must be used, as
+`cc-options` is *not* sufficient.
+
+This issue affects any project that uses `hs-bindgen`, even if it is used as a
+transitive dependency.
+
 ### Common errors and solutions
 
 * Missing headers (`stddef.h`, etc.): If you encounter errors about missing

